File: 2019/01/rs/src/main.rs
use std::env;
use std::fs;
use std::time::{Instant};

macro_rules! mass_fuel {
    ($mass:expr) => {
        $mass / 3 - 2
    };
}

fn get_fuel(mass: &i32) -> i32 {
    let mut total = 0;
    let mut current_mass = *mass;
    loop {
        let fuel: i32 = mass_fuel!(current_mass as i32);
        if fuel <= 0 {
            return total
        }
        total += fuel;
        current_mass = fuel;
    }
}

fn solve(masses: &Vec<i32>) -> (i32, i32) {
    (
        masses.iter().map(|mass| mass_fuel!(mass)).sum(), 
        masses.iter().map(get_fuel).sum()
    )
}

fn get_input(file_path: &String) -> Vec<i32> {
    fs::read_to_string(file_path)
        .expect("Error reading input file!")
        .split("\n").filter(|&line| !line.is_empty())
        .map(|line| line.parse().expect(&format!("Unable to parse '{}'", line)))
        .collect()
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        panic!("Please, add input file path as parameter");
    }
    let now = Instant::now();
    let (part1_result, part2_result) = solve(&get_input(&args[1]));
    println!("P1: {}", part1_result);
    println!("P1: {}", part2_result);
    println!();
    println!("Time: {:?}", now.elapsed().as_secs_f32());
}

File: 2019/01/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>

typedef struct
{
    int *masses;
    int count;
} Input;
typedef struct
{
    int part1;
    int part2;
} Results;

int part1(int mass)
{
    return mass / 3 - 2;
}

int part2(int mass)
{
    int total = 0;
    int currentMass = mass;
    while (1)
    {
        int fuel = currentMass / 3 - 2;
        if (fuel <= 0)
            return total;
        total += fuel;
        currentMass = fuel;
    }
}

int getTotal(Input input, int (*fuelCalculator)(int))
{
    int total = 0;
    while (input.count--)
        total += (*fuelCalculator)(*input.masses++);
    return total;
}

Results solve(Input input)
{
    return (Results){getTotal(input, &part1), getTotal(input, &part2)};
}

Input getInput(char *filePath)
{
    FILE *file;
    if ((file = fopen(filePath, "r")) == NULL)
    {
        perror("Error reading input file!\n");
        exit(1);
    }
    char *line = NULL;
    size_t len = 0;
    size_t size = 16;
    int count = 0;
    int *masses = malloc(128 * sizeof(int));
    int *current = masses;
    while (getline(&line, &len, file) != EOF)
    {
        count++;
        *current = atoi(line);
        current++;
    }
    if (line)
        free(line);
    fclose(file);
    return (Input){masses, count};
}

void freeInput(Input input)
{
}

int main(int argc, char **argv)
{
    if (argc != 2)
    {
        perror("Please, add input file path as parameter");
        exit(1);
    }
    struct timeval starts, ends;
    gettimeofday(&starts, NULL);
    Input input = getInput(argv[1]);
    Results results = solve(input);
    gettimeofday(&ends, NULL);
    freeInput(input);
    printf("P1: %d\n", results.part1);
    printf("P2: %d\n\n", results.part2);
    printf("Time: %.7f\n", (double)((ends.tv_sec - starts.tv_sec) * 1000000 + ends.tv_usec - starts.tv_usec) / 1000000);
    return 0;
}